From 6beb08c32c3c9cd393b73afa58720eae69a37db3 Mon Sep 17 00:00:00 2001
From: 916BGAI <zhangjian@sipeed.com>
Date: Mon, 3 Mar 2025 09:48:34 +0800
Subject: [PATCH 2/2] fix emmc boot fail

---
 drivers/mmc/sunxi_mmc.c | 5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

diff --git a/drivers/mmc/sunxi_mmc.c b/drivers/mmc/sunxi_mmc.c
index 074b08ac5d8..8621d6cbcbd 100644
--- a/drivers/mmc/sunxi_mmc.c
+++ b/drivers/mmc/sunxi_mmc.c
@@ -27,6 +27,7 @@
 #include <asm/arch/cpu.h>
 #include <asm/arch/mmc.h>
 #include <linux/delay.h>
+#include <stdio.h>
 #include <sunxi_gpio.h>
 
 #ifndef CCM_MMC_CTRL_MODE_SEL_NEW
@@ -541,7 +542,7 @@ struct mmc *sunxi_mmc_init(int sdc_no)
 	cfg->f_max = 52000000;
 
 	if(sdc_no==2) {
-		cfg->f_max = 20000000;
+		cfg->f_max = 8000000;
 	}
 
 	if (mmc_resource_init(sdc_no) != 0)
@@ -657,7 +658,7 @@ static int sunxi_mmc_probe(struct udevice *dev)
 		return ret;
 
 	if (bus_width == 8) {
-		cfg->f_max = 20000000;
+		cfg->f_max = 8000000;
 	}
 
 	priv->reg = dev_read_addr_ptr(dev);
-- 
2.48.1

